### ISSUE TITLE
No superclass mismatch when class is private

### ISSUE THREAD

**Issue by jwoertink** [crystal-lang/crystal#9646]
I'm not sure if this is a bug or intended, but if you have this code:

```crystal
class Base
end
 
class Base < Base
end
```

You'll get the error https://play.crystal-lang.org/#/r/9h18
> Error: superclass mismatch for class Base (Base for Reference)

But if you make it private, then it will compile fine. https://play.crystal-lang.org/#/r/9h19

```crystal
class Base
end

private class Base < Base
end
```

Maybe because `private class Base` != `class Base` ?

**Comment by asterite** [comment]
Nice! I tried the same thing in Ruby, for the first example, and I get the same error. It seems we copied Ruby very well :-D

The second example is fine. When you define `private class Base` that basically makes `Base` in the file refer to the private one. But before that happens the superclass is checked, and at that point `Base` is the non-private one.

It goes without saying that you shouldn't do this unless you want to confuse other people or your future self :-)

**Comment by jwoertink** [comment]
Yeah, we accidentally had that [written in a spec](https://github.com/luckyframework/lucky/blob/master/spec/lucky/action_route_params_spec.cr#L5), and I was confused on how it even worked 😅 . I tried it and saw it threw an error (like I expected), then noticed that it didn't with the private.

If that's intended behavior, then I'm ok with closing this out.

**Comment by asterite** [comment]
I think it's actually "very nice". You can replace a well-known type with a different version of it in a spec file, like in your example. However, I admit it can be a bit confusing...
